src/CoreFn/Expr.purs

```diff
@@ -131,69 +52,112 @@ data Expr a
   -- Variable
   --
   | Var a (Qualified Ident)
+  -- |
+  -- A case expression
+  --
+  | Case a (Array (Expr a)) (Array (CaseAlternative a))
+  -- |
+  -- A let binding
+  --
+  | Let a (Array (Bind a)) (Expr a)
 
 derive instance eqExpr :: Eq a => Eq (Expr a)
+derive instance functorExpr :: Functor Expr
 derive instance ordExpr :: Ord a => Ord (Expr a)
 
 instance showExpr :: Show a => Show (Expr a) where
-  show (Literal x y) = "(Literal " <> show x <> " " <> show y <> ")"
-  show (Abs x y z) = "(Abs " <> show x <> " " <> show y <> " " <> show z <> ")"
-  show (App x y z) = "(App " <> show x <> " " <> show y <> " " <> show z <> ")"
-  show (Var x y) = "(Var " <> show x <> " " <> show y <> ")"
-
-readExpr :: Foreign -> F (Expr Unit)
-readExpr x = do
-  label <- index x 0 >>= readString
-  readExpr' label x
-
-  where
-
-  readExpr' :: String -> Foreign -> F (Expr Unit)
-  readExpr' "Literal" y = do
-    value <- index y 1
-    Literal unit <$> readLiteral value
-  readExpr' "Abs" y = do
-    ident <- index y 1
-    expr <- index y 2
-    Abs unit <$> readIdent ident <*> readExpr expr
-  readExpr' "App" y = do
-    expr1 <- index y 1
-    expr2 <- index y 2
-    App unit <$> readExpr expr1 <*> readExpr expr2
-  readExpr' "Var" y = do
-    value <- index y 1
-    Var unit <$> readQualified Ident value
-  readExpr' label _ = fail $ ForeignError $ "Unknown expression: " <> label
-
-readExprJSON :: String -> F (Expr Unit)
-readExprJSON = parseJSON >=> readExpr
+  show (Literal a l) =
+    "(Literal " <>
+      intercalate " " [ show a, show l ] <>
+    ")"
+  show (Constructor a t c fs) =
+    "(Constructor " <>
+      intercalate " " [ show a, show t, show c, show fs] <>
+    ")"
+  show (Accessor a s e) =
+    "(Accessor " <>
+      intercalate " " [ show a, show s, show e ] <>
+    ")"
+  show (ObjectUpdate a e fs) =
+    "(ObjectUpdate " <>
+      intercalate " " [ show a, show e, show fs ] <>
+    ")"
+  show (Abs a i e) =
+    "(Abs " <>
+      intercalate " " [ show a, show i, show e ] <>
+    ")"
+  show (App a e1 e2) =
+    "(App " <>
+      intercalate " " [ show a, show e1, show e2 ] <>
+    ")"
+  show (Var a q) =
+    "(Var " <>
+      intercalate " " [ show a, show q ] <>
+    ")"
+  show (Case a es cs) =
+    "(Case " <>
+      intercalate " " [ show a, show es, show cs ] <>
+    ")"
+  show (Let a bs e) =
+    "(Let " <>
+      intercalate " " [ show a, show bs, show e ] <>
+    ")"
+
 
 -- |
 --  A let or module binding.
 --
-data Bind a = Bind (Array (Tuple (Tuple a Ident) (Expr a)))
+data Bind a
+  = NonRec a Ident (Expr a)
+  | Rec (Array (Tuple (Tuple a Ident) (Expr a)))
 
 derive instance eqBind :: Eq a => Eq (Bind a)
 derive instance ordBind :: Ord a => Ord (Bind a)
 
+instance functorBindRec :: Functor Bind where
+  map f (NonRec a i e) = NonRec (f a) i (map f e)
+  map f (Rec ts) = Rec $ map (bimap (lmap f) (map f)) ts
```

Comment:
@liamgoodacre does this look right to you?

FWIW, `stack build --ghc-options "-ddump-deriv"` produced `.stack-work/dist/x86_64-osx/Cabal-2.0.0.2/build/src/Language/PureScript/CoreFn/Expr.dump-deriv` which contained this:

```
instance GHC.Base.Functor
             Language.PureScript.CoreFn.Expr.Bind where
  GHC.Base.fmap
    f_amsj
    (Language.PureScript.CoreFn.Expr.NonRec a1_amsk a2_amsl a3_amsm)
    (Language.PureScript.CoreFn.Expr.NonRec a1_amsk a2_amsl a3_amsm)
    = Language.PureScript.CoreFn.Expr.NonRec
        (f_amsj a1_amsk)
        ((\ b1_amsn -> b1_amsn) a2_amsl)
        (GHC.Base.fmap f_amsj a3_amsm)
  GHC.Base.fmap f_amso (Language.PureScript.CoreFn.Expr.Rec a1_amsp)
    = Language.PureScript.CoreFn.Expr.Rec
        (GHC.Base.fmap
            (\ b3_amsq
              -> case b3_amsq of {
                    ((,) a1_amsr a2_amss)
                      -> (,)
                          ((\ b2_amst
                              -> case b2_amst of {
                                    ((,) a1_amsu a2_amsv)
                                      -> (,) (f_amso a1_amsu) ((\ b1_amsw -> b1_amsw) a2_amsv) })
                              a1_amsr)
                          (GHC.Base.fmap f_amso a2_amss) })
            a1_amsp)
```